### PR TITLE
Implemented ScheduleMessageAsync and CancelScheduledMessageAsync of messages

### PR DESCRIPTION
**Pull request by nemakam**
## Description

Exposed the following two methods:
1. Task<long> ScheduleMessageAsync(BrokeredMessage message, DateTimeOffset scheduleEnqueueTimeUtc)
2. Task CancelScheduledMessageAsync(long sequenceNumber)

They are exposed in both QueueClient and TopicClient and can be used to handle scheduling of messages to be delivered in the future.

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] **I have read the [contribution guidelines](./CONTRIBUTING.md).**
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR.
- [x] The pull request does not introduce breaking changes (unless a major version change occurs in the assembly and module).
- [x] If applicable, the public code is properly documented.
- [x] Pull request includes test coverage for the included changes.
- [x] The code builds without any errors.

### CODE REVIEW

**Comment by SeanFeldman** [comment]
Is this implementing #55 ? Would be nice to indicate what issue it implements/resolves in the general description 😃

File: test/Microsoft.Azure.ServiceBus.UnitTests/SenderReceiverClientTestBase.cs [hunk 1]
```diff
@@ -157,5 +158,60 @@ public abstract class SenderReceiverClientTestBase
 
             var receivedMessages = await TestUtility.ReceiveMessagesAsync(messageReceiver, messageCount);
         }
+
+        protected async Task ScheduleMessagesAsyncTestCase(MessageSender messageSender, MessageReceiver messageReceiver, int messageCount)
```
Comment (SeanFeldman) [inline test/Microsoft.Azure.ServiceBus.UnitTests/SenderReceiverClientTestBase.cs]: Would be really nice to see a test name saying "scheduled message appears on or after the time it was schedule for".

**Comment by nemakam** [comment]
Thanks Sean. Yes, #55 will be resolved after this PR. :)

File: test/Microsoft.Azure.ServiceBus.UnitTests/SenderReceiverClientTestBase.cs [hunk 2]
```diff
@@ -157,5 +158,60 @@ public abstract class SenderReceiverClientTestBase
...
+
+            var sequenceNumber = await messageSender.ScheduleMessageAsync(brokeredMessage, scheduleTime);
+            TestUtility.Log($"Received sequence number: {sequenceNumber}");
+            Assert.True(sequenceNumber > 0);
+
+            TestUtility.Log("Cancelling scheduled message");
+            await messageSender.CancelScheduledMessageAsync(sequenceNumber);
+
+            TestUtility.Log("Sleeping for 20 seconds...");
+            Thread.Sleep(TimeSpan.FromSeconds(20));
```
Comment (SeanFeldman) [inline test/Microsoft.Azure.ServiceBus.UnitTests/SenderReceiverClientTestBase.cs]: `await Task.Delay()` instead?

File: src/Microsoft.Azure.ServiceBus/Amqp/AmqpMessageSender.cs [hunk 3]
```diff
@@ -65,6 +86,75 @@ public override Task CloseAsync()
...
+                request.Map[ManagementConstants.Properties.Messages] = new List<AmqpMap> { entry };
+
+                IEnumerable<long> sequenceNumbers = null;
+                var response = await this.ExecuteRequestResponseAsync(request);
+                if (response.StatusCode == AmqpResponseStatusCode.OK)
+                {
+                    sequenceNumbers = response.GetValue<long[]>(ManagementConstants.Properties.SequenceNumbers);
+                }
+
+                return sequenceNumbers?.FirstOrDefault() ?? 0;
```
Comment (SeanFeldman) [inline src/Microsoft.Azure.ServiceBus/Amqp/AmqpMessageSender.cs]: So "0" is a valid sequence number if nothing was returned from the server?

File: test/Microsoft.Azure.ServiceBus.UnitTests/SenderReceiverClientTestBase.cs [hunk 4]
```diff
@@ -157,5 +158,60 @@ public abstract class SenderReceiverClientTestBase
...
+        protected async Task CancelScheduledMessagesAsyncTestCase(MessageSender messageSender, MessageReceiver messageReceiver, int messageCount)
+        {
+            var scheduleTime = new DateTimeOffset(DateTime.UtcNow).AddSeconds(20);
+            var brokeredMessage = new BrokeredMessage("Test1") { MessageId = Guid.NewGuid().ToString() };
+            TestUtility.Log(
+                $"Sending message with schedule time: {scheduleTime.UtcDateTime} and messageID {brokeredMessage.MessageId}");
+
+            var sequenceNumber = await messageSender.ScheduleMessageAsync(brokeredMessage, scheduleTime);
+            TestUtility.Log($"Received sequence number: {sequenceNumber}");
+            Assert.True(sequenceNumber > 0);
```
Comment (SeanFeldman) [inline test/Microsoft.Azure.ServiceBus.UnitTests/SenderReceiverClientTestBase.cs]: Is this a redundant assertion? based on the [Enqueue code](https://github.com/Azure/azure-service-bus-dotnet/pull/69/files#diff-d2db9464d2fb826ab51d7b797c9ceef6R134), sequence number can be zero.
Plus, we'll assert that this message is not found by [follow up assertions](https://github.com/Azure/azure-service-bus-dotnet/pull/69/files#diff-d330d38fd2ca3f843142501fc38c00c7R213).

**Comment by SeanFeldman** [comment]
@nemakam just a suggestion - to have in the general description (first comment) reference to what work item is covered (issue). That was it's trackable on the issue and easy to see when PR is merged that work is completed. Example how it would like on the issue:
image

File: src/Microsoft.Azure.ServiceBus/Amqp/AmqpMessageSender.cs [hunk 5]
```diff
@@ -65,6 +86,75 @@ public override Task CloseAsync()
             }
         }
 
+        protected override async Task<long> OnScheduleMessageAsync(BrokeredMessage brokeredMessage)
+        {
+            // TODO: Ensure System.Transactions.Transaction.Current is null. Transactions are not supported by 1.0.0 version of dotnet core.
+            TimeoutHelper timeoutHelper = new TimeoutHelper(this.OperationTimeout, true);
+            using (AmqpMessage amqpMessage = AmqpMessageConverter.ClientGetMessage(brokeredMessage))
+            {
+                RequestResponseAmqpLink amqpLink =
```
Comment (nemakam) [inline src/Microsoft.Azure.ServiceBus/Amqp/AmqpMessageSender.cs]: Its not needed. That's a cleanup that was missed out. Thanks

File: src/Microsoft.Azure.ServiceBus/Amqp/AmqpMessageSender.cs [hunk 6]
```diff
@@ -65,6 +86,75 @@ public override Task CloseAsync()
...
+                    ManagementConstants.Operations.ScheduleMessageOperation,
+                    this.OperationTimeout,
+                    null);
+
+                ArraySegment<byte>[] payload = amqpMessage.GetPayload();
+                BufferListStream buffer = new BufferListStream(payload);
+                ArraySegment<byte> value = buffer.ReadBytes((int)buffer.Length);
+
+                var entry = new AmqpMap();
+                {
```
Comment (nemakam) [inline src/Microsoft.Azure.ServiceBus/Amqp/AmqpMessageSender.cs]: The API on the service side (broker) has a slightly different implementation for ScheduleMessageAsync(). These few lines are required just to take care of that. They will be updated when we update the service side code. 
Not fixing this for now.

File: src/Microsoft.Azure.ServiceBus/Amqp/AmqpMessageSender.cs (hunk 3 above)
Comment (nemakam) [inline src/Microsoft.Azure.ServiceBus/Amqp/AmqpMessageSender.cs]: 0 is not a valid sequence number. I've written it as such to keep it consistent with the other published library. I'm working on a task for exception handling in this path. Will send out a separate PR for that... Leaving it as is for this PR.

File: src/Microsoft.Azure.ServiceBus/Amqp/AmqpMessageSender.cs [hunk 7]
```diff
@@ -65,6 +86,75 @@ public override Task CloseAsync()
...
+                return sequenceNumbers?.FirstOrDefault() ?? 0;
+            }
+        }
+
+        protected override async Task OnCancelScheduledMessageAsync(long sequenceNumber)
+        {
+            // TODO: Ensure System.Transactions.Transaction.Current is null. Transactions are not supported by 1.0.0 version of dotnet core.
+            TimeoutHelper timeoutHelper = new TimeoutHelper(this.OperationTimeout, true);
+
+            RequestResponseAmqpLink amqpLink =
```
Comment (nemakam) [inline src/Microsoft.Azure.ServiceBus/Amqp/AmqpMessageSender.cs]: Removed it...

File: src/Microsoft.Azure.ServiceBus/MessageSender.cs [hunk 8]
```diff
@@ -35,18 +35,55 @@ public Task SendAsync(IEnumerable<BrokeredMessage> brokeredMessages)
...
+                ValidateMessage(brokeredMessage);
+            }
+        }
+
+        static void ValidateMessage(BrokeredMessage brokeredMessage)
+        {
+            if (brokeredMessage.IsLockTokenSet)
             {
-                throw Fx.Exception.Argument(nameof(brokeredMessages), "Cannot Send ReceivedMessages");
+                throw Fx.Exception.Argument(nameof(brokeredMessage), "Cannot Send ReceivedMessages");
```
Comment (nemakam) [inline src/Microsoft.Azure.ServiceBus/MessageSender.cs]: The `validateMessage()` is being called only for `SendMessage` and `ScheduleMessage` which are pretty much same.

File: src/Microsoft.Azure.ServiceBus/MessageSender.cs [hunk 9]
```diff
@@ -35,18 +35,55 @@ public Task SendAsync(IEnumerable<BrokeredMessage> brokeredMessages)
...
+            if (scheduleEnqueueTimeUtc.CompareTo(DateTimeOffset.UtcNow) < 0)
+            {
+                throw Fx.Exception.ArgumentOutOfRange(
+                    nameof(scheduleEnqueueTimeUtc),
+                    scheduleEnqueueTimeUtc.ToString(),
+                    "Cannot schedule messages in the past");
+            }
+
+            message.ScheduledEnqueueTimeUtc = scheduleEnqueueTimeUtc.UtcDateTime;
+            MessageSender.ValidateMessage(message);
```
Comment (nemakam) [inline src/Microsoft.Azure.ServiceBus/MessageSender.cs]: Updated it to 'Cannot send a message that was already received.'
Keeping it generic wrt send / schedule.

File: test/Microsoft.Azure.ServiceBus.UnitTests/SenderReceiverClientTestBase.cs (hunk 1 above)
Comment (nemakam) [inline test/Microsoft.Azure.ServiceBus.UnitTests/SenderReceiverClientTestBase.cs]: Updated the test name.

File: test/Microsoft.Azure.ServiceBus.UnitTests/SenderReceiverClientTestBase.cs (hunk 4 above)
Comment (nemakam) [inline test/Microsoft.Azure.ServiceBus.UnitTests/SenderReceiverClientTestBase.cs]: Sequence number cannot be 0. The check for greater than 0 exists to ensure that message was indeed scheduled without error. Refer to previous reply that I just gave on the other comment

File: src/Microsoft.Azure.ServiceBus/MessageSender.cs (hunk 9 above)
Comment (SeanFeldman) [inline src/Microsoft.Azure.ServiceBus/MessageSender.cs]: 👍

File: test/Microsoft.Azure.ServiceBus.UnitTests/SenderReceiverClientTestBase.cs (hunk 4 above)
Comment (SeanFeldman) [inline test/Microsoft.Azure.ServiceBus.UnitTests/SenderReceiverClientTestBase.cs]: Ok. Based on another reply there would be an exception thrown, which leads me to think this will change (removed check for sequence number)
